src/randgraphs_sbm.jl

```diff
@@ -0,0 +1,185 @@
+    B = diagm(internalp) + externalp*(ones(numblocks, numblocks)-I)
+    return B
+end
+
+function StochasticBlockModel(internalp::Float64,
+                              externalp::Float64,
+                              size::Int,
+                              numblocks::Int)
+    sizes = [size for i in 1:numblocks]
+    B = sbmaffinity([internalp for i in 1:numblocks], externalp, sizes)
+    StochasticBlockModel(sizes, B)
+end
+
+function StochasticBlockModel(internalp::Vector{Float64}, externalp::Float64, sizes::Vector{Int})
+    B = sbmaffinity(internalp, externalp, sizes)
+    return StochasticBlockModel(sizes, B)
+end
+
+
+biclique = ones(2,2) - eye(2)
+
+"""construct the affinity matrix for a near bipartite SBM.
+between is the affinity between the two parts of each bipartite community
+intra is the probability of an edge within the parts of the partitions.
+
+This is a specific type of SBM with k/2 blocks each with two halves.
+Each half is connected as a random bipartite graph with probability `intra`
+The blocks are connected with probability `between`.
+"""
+function nearbipartiteaffinity(sizes::Vector{Int}, between::Float64, intra::Float64)
+    numblocks = round(Int, length(sizes)/2)
+    return kron(between*eye(numblocks), biclique) + eye(2numblocks)*intra
+end
+
+"""Return a generator for edges from a stochastic block model near bipartite graph."""
+function nearbipartiteaffinity(sizes::Vector{Int}, between::Float64, inter::Float64, noise)
+    B = nearbipartiteaffinity(sizes, between, inter) + noise
+    #= info("Affinities are:\n$B")#, file=stderr) =#
+    return B
+end
+
+function nearbipartiteSBM(sizes, between, inter, noise)
+    return StochasticBlockModel(sizes, nearbipartiteaffinity(sizes, between, inter, noise))
+end
+
+
+"""random_pair: generates a stream of random pairs in 1:n"""
+function random_pair(n::Int)
+    while true
+        produce( rand(1:n), rand(1:n) )
+    end
+end
+
+"""take the first n elements produced by t"""
+function take(n::Int, t::Task) 
+    for i in 1:n
+        produce(consume(t))
+    end
+end
+
+""" Draw an edgestream
+Affinity::(Node, Node) -> Probability
+pairs::generator of (Node, Node)
+"""
+function sample(affinity, pairs)
+    for (i, j) in pairs
+        aff = affinity(i,j)
+        #print(STDERR, "affinity $i,$j = $aff\n")
+        if rand() < aff
+            produce((i,j))
+        #else try again
+        end        
+    end
+end
+
+"""Take an infinite sample from the sbm. Use take() to take only the first n edges
+or pass to graph(edgestream, numedges).
+"""
+function sample(sbm::StochasticBlockModel)
+    pairs = @task random_pair(sbm.n)
+    for (i,j) in pairs
+    	if i == j
+	    continue
+	end
+        p = sbm.affinities[sbm.nodemap[i], sbm.nodemap[j]]
+        if rand() < p
+            produce(i, j)
+        end
+    end
+end
+
+"""convert a stream of edges produced by sample into a graph"""
+function graph(edgestream, sizes, numedges::Int)
```

Comment:
Not that I'm aware of. The closest I can come is `eltype` but that can't be specified within the parameter list.

Not to bikeshed, but can we change the name of the function to something more descriptive? Not this, necessarily, but I'm thinking of something like `graph_from_edgestream()` - just so that we don't have confusion with the constructor.